daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/EvTextNumber.scala

```diff
@@ -51,6 +66,163 @@ class TextStandardExponentRepEv(expr: CompiledExpression[String], tci: DPathComp
   override lazy val runtimeDependencies = Vector()
 }
 
+class TextNumberFormatEv(
+  tci: DPathCompileInfo,
+  decimalSepEv: Maybe[TextStandardDecimalSeparatorEv],
+  groupingSepEv: Maybe[TextStandardGroupingSeparatorEv],
+  exponentRepEv: Maybe[TextStandardExponentRepEv],
+  infRep: Maybe[String],
+  nanRep: Maybe[String],
+  checkPolicy: TextNumberCheckPolicy,
+  textNumberPattern: String,
+  rounding: TextNumberRounding,
+  roundingMode: Maybe[TextNumberRoundingMode],
+  roundingIncrement: MaybeDouble,
+  zeroRepsRaw: List[String],
+  primType: PrimType)
+  extends Evaluatable[ThreadLocal[DecimalFormat]](tci)
+  with InfosetCachedEvaluatable[ThreadLocal[DecimalFormat]] {
```

Comment:
Since this is an evaluatable, this threadlocal can just be a member of the state now easily. Since states are per-thread anyway that is.

Reply:
I thought if the Evaluatable is constant (which is probably the common case here since text number properties tend to not use expressions) then the result would be be computed only once at compile time and shared among threads? Or maybe I'm misunderstanding what you're suggesting?

Reply:
I just saw evaluatable parameterized by thread-local type, and that seems incongruous to me. A thread-local is really a hack that uses the current thread ID and a hash table to find a thread-specific instance. We don't need to use that mechanism since the state is a thread-specific object. 

So a thread-local and an evaluatable should never go together.

Even if everything folds away to a constant you still sometimes need the DecimalFormat object at runtime (I think.... )

Reply:
My understanding is that there is single Evaluatable instance (e.g. TextNumberFormatEv) that is passed to a parser, and this Evaluatable is shared among threads. There is a "maybeConst" member variable in the Evaluatable that stores the compile time result of the Evaluatable if it is constant. In that case, that constant value will be shared between different threads when evaluate(state) is called. The state is never used when it's constant.

In this particular case, DecimalFormat isn't thread-safe, so we need to have a unique instance per-thread, so the ThreadLocal is used for that. If the Evaluatable isn't constant, then we don't technically' need the ThreadLocal, since we'll get a new DecimalFormat everytime time  evaluate(state) is called, but I'm not sure there's a good way to handle a ThreadLocal[DecimalFormat] in once case or a DecimalFormat in another.

One option would be to never cache the result and always allocate a new DecimalFormat everytime evaluate is called, but the creation of DecimalFormat is probably pretty slow since it needs to parse text number patterns and do all kinds of other stuff.

Perhaps we really need an EvaluatableNonThreadSafe, which uses a ThreadLocal behind the scenes for when it's a constant?

Reply:
Pretty sure it's way simpler than this.

Here's the key observation: The infoset is part of the state. Caching on the infoset is a private thread-safe cache. The Evaluatable is taking care of that for you.

So I believe you can just always allocate a DecimalFormat as the result of the compute method.

Reply:
Doesn't that mean a new DecimalFormat will be allocated and initialized for every parse? That can be expensive. As I understand it, the purpose of the const member in Evaluatable is if something can be allocated and initialized at schema compile time then we do it then, avoiding that overhead during parse time. The consequene of this though, is that different threads now share that same allocated and initialized constant, which in the DecimalFormat case makes things not thread-safe.

Reply:
Ah. I was worried about the case where it's non-constant. That's not the problem. 

I think if you have an evaluatable, that if it produces a constant value, but..... that constant is actually mutable non-thread-safe instance, ....it's a bit strange.

But your techinque works fine for this, and ThreadLocal.get is very tight code that doesn't synchronize or anything,  so this is fine. 

Doesn't ThreadLocal have to be Serializable in this case though? I don't think it is from looking at Javadoc,  but how could your code be working if it isn't? It's trivial for it to be serializable though, so maybe it is, and just isn't mentioned in javadoc?

There is another analogous place which is the way Charset Encoder/Decoder work. Those are stateful non-thread safe, but potentially if encoding is known they are these quasi-constants in that they are needed to be per-thread instances. There's a CharsetEv, but not EncoderEv nor DecoderEv. Instead there's a cache for these maintained in the State. There's a bunch of other stuff going on there to deal with fast detection that since the charset didn't change the Encoder/Decoder can't change, but this techique using ThreadLocal still may be a simplifier as well there.

Reply:
Yeah, it's definitely feels strange.

And yep, ThreadLocal does need to be Serializable but it's not. I found that if I just mixed Serializable in everything worked as expected. This is somewhere in this patch:
```scala
val res = new ThreadLocal[DecimalFormat] with Serializable { ... }
```
I think maybe the difference between the charset cache and this is there's a simple way to key on an already allcoated charset encoder/decoder--it's just the name of the charset. But with decimal format, there isn't a simple key--it takes a bunch of different parameters.

That said, that sort of cache probably makes a ton of sense. Very few schemas are going to have more than a couple decimal formats, so having a unique one for every since text number parser is a pretty big waste of memory. I think that's a bit out of scope for this refactoring though (this is basically the same thing we had before, just using Evaluatables instead of CacheDynamic). I'll create a bug to think about how we can implement this to reduce the number of DecimalFormats allocated.